test/hotspot/jtreg/runtime/FieldLayout/ArrayBaseOffsets.java

```diff
@@ -0,0 +1,113 @@
+ * @run main/othervm -XX:+UseCompressedOops -XX:+UseCompressedClassPointers ArrayBaseOffsets
+ */
+/*
+ * @test id=no-coops-no-ccp
+ * @library /test/lib
+ * @requires vm.bits == "64"
+ * @modules java.base/jdk.internal.misc
+ * @run main/othervm -XX:-UseCompressedOops -XX:-UseCompressedClassPointers ArrayBaseOffsets
+ */
+/*
+ * @test id=no-coops-with-ccp
+ * @library /test/lib
+ * @requires vm.bits == "64"
+ * @requires vm.opt.UseCompressedClassPointers != false
+ * @modules java.base/jdk.internal.misc
+ * @run main/othervm -XX:-UseCompressedOops -XX:+UseCompressedClassPointers ArrayBaseOffsets
+ */
+/*
+ * @test id=32bit
+ * @library /test/lib
+ * @requires vm.bits == "32"
+ * @modules java.base/jdk.internal.misc
+ * @run main/othervm ArrayBaseOffsets
+ */
+
+import jdk.internal.misc.Unsafe;
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.RuntimeMXBean;
+import java.util.List;
+
+import jdk.test.lib.Asserts;
+import jdk.test.lib.Platform;
+
+public class ArrayBaseOffsets {
+
+    private static final boolean COOP;
+    private static final boolean CCP;
+
+    static {
+        if (Platform.is64bit()) {
+            RuntimeMXBean runtime = ManagementFactory.getRuntimeMXBean();
+            List<String> vmargs = runtime.getInputArguments();
+            CCP = !vmargs.contains("-XX:-UseCompressedClassPointers");
+            COOP = System.getProperty("java.vm.compressedOopsMode") != null;
+        } else {
+            COOP = CCP = false;
+        }
+    }
+
+    static public void main(String[] args) {
+        Unsafe unsafe = Unsafe.getUnsafe();
+        int intOffset, longOffset;
+        if (Platform.is64bit()) {
+            if (CCP) {
+                intOffset = 16;
+                longOffset = 16;
+            } else {
+                intOffset = 20;
+                longOffset = 24;
+            }
+        } else {
+            intOffset = 12;
+            longOffset = 16;
+        }
+        Asserts.assertEquals(unsafe.arrayBaseOffset(boolean[].class), intOffset,  "Misplaced boolean array base");
+        Asserts.assertEquals(unsafe.arrayBaseOffset(byte[].class),    intOffset,  "Misplaced byte    array base");
+        Asserts.assertEquals(unsafe.arrayBaseOffset(char[].class),    intOffset,  "Misplaced char    array base");
+        Asserts.assertEquals(unsafe.arrayBaseOffset(short[].class),   intOffset,  "Misplaced short   array base");
+        Asserts.assertEquals(unsafe.arrayBaseOffset(int[].class),     intOffset,  "Misplaced int     array base");
+        Asserts.assertEquals(unsafe.arrayBaseOffset(long[].class),    longOffset, "Misplaced long    array base");
+        Asserts.assertEquals(unsafe.arrayBaseOffset(float[].class),   intOffset,  "Misplaced float   array base");
+        Asserts.assertEquals(unsafe.arrayBaseOffset(double[].class),  longOffset, "Misplaced double  array base");
+        int expectedObjArrayOffset = (COOP || !Platform.is64bit()) ? intOffset : longOffset;
```

Comment:
`|| !Platform.is64Bit()` looks redundant, since `COOP` is guaranteed to be false on 32-bit?

Reply:
No, because on 32 bit this expression becomes true, which is what we want - the small (int) offset for reference arrays.

Reply:
Right, OK.